power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/observability.hpp

```diff
@@ -58,29 +28,114 @@ std::tuple<Idx, Idx> count_voltage_sensors(const Idx n_bus, const MeasuredValues
     return std::make_tuple(n_voltage_sensor, n_voltage_phasor_sensor);
 }
 
-} // namespace detail
+// count flow sensors into ybus structure like
+// lower triangle part is always zero
+// for diagonal part, it will be one if there is bus injection
+// for upper triangle part, it will be one if there is branch flow sensor and the branch is fully connected
 template <symmetry_tag sym>
-inline void necessary_observability_check(MeasuredValues<sym> const& measured_values,
-                                          std::shared_ptr<MathModelTopology const> const& topo) {
+std::vector<int8_t> count_flow_sensors(MeasuredValues<sym> const& measured_values, MathModelTopology const& topo,
+                                       YBusStructure const& y_bus_structure) {
+    Idx const n_bus{topo.n_bus()};
+    std::vector<int8_t> flow_sensors(y_bus_structure.row_indptr.back(), 0); // initialize all to zero
```

Comment:
Since `flow_sensors` will be 1's and 0's, why not make it `std::vector<bool>`. Since this state for later use is not accumulable, booleans seems perfect.

Reply:
If we prefer to have `int_8`, we could make it one bigger, and use the final entry to count along the way. This avoids the `O(n)` check for `n_flow_sensors`.

Reply:
`std::vector<bool>` is not STL container. So I do not use them.

Count along the way will not help for the sufficient condition check. In the second step we need to move things around and re-count.

Reply:
Apart from not continuous in memory, `std::vector<bool>` is the same the way a container is used here?

Reply:
this is not actually a bool. See my other comments: it is a meaningful number: number of independent sensors.